Resolve SME feedback in best practices (#347)
osisoft/OCS-Docs: Content_Portal/How_To_Guides/Add-Org/clients.md

> # Clients

Reviewer: @osisoft/product-readiness-guild Please review

## Definition

Clients provide a way for users or applications to authenticate against OSIsoft Cloud Services (OCS) from applications other than the OCS portal itself. OCS supports three types of clients, intended to support different types of applications.

You must have the **Account Administrator** role to add and manage clients in a tenant.

## Client-Credentials

> Client-credentials clients are used for server-to-server communication where no user interaction is required. The client typically authenticates with the token endpoint using its client ID and secret. A secret is a unique key generated for each client, to connect to OSIsoft assets, resources and services for a limited period of time.

Author: ```suggestion
Client-credentials clients are used for server-to-server communication where no user interaction is required. The client typically authenticates with the token endpoint using its client ID and secret. A secret is a unique key generated for each client to connect to OSIsoft assets, resources, and services for a time-limited period.
```

### Client-Credentials PI Core Counterpart

> Client-credentials clients are much like Windows service accounts that might be used to authenticate against the PI Data Archive or PI AF Server, and it is thus very important to keep the secret secure in much the same way it is important to keep service account passwords secure.

Author: ```suggestion
Client-credentials clients are very similar to Windows service accounts that might be used to authenticate against the PI Data Archive or PI AF Server.  Therefore, it is very important to keep secrets secure in the same way that it is important to keep service account passwords secure.
```

> Client-credentials clients are very similar to Windows service accounts that might be used to authenticate against the PI Data Archive or PI AF Server. Therefore, it is very important to keep secrets secure in the same way that it is important to keep service account passwords secure.

Author: ```suggestion
Client-credentials clients are very similar to Windows service accounts that might be used to authenticate against PI Data Archive or PI AF server. Therefore, it is very important to keep secrets secure in the same way that it is important to keep service account passwords secure.
```

> ### Client-Credentials Best Practices

Author: ```suggestion
### Client-credentials best practices
```

> 1. Create a separate Client-credentials client for each separate device or instance of an application that will connect to to OCS. This will ensure that it is possible to manage secrets discretely for individual applications and to know which applications are connecting to OCS.

Author: ```suggestion
1. Create a separate client-credentials client for each device or instance of an application that connects to OCS. This ensures that secrets can be discretely managed for individual applications and that you know which applications are connecting to OCS.
```

1. Ensure that client secrets are stored securely where they are used.

> 1. Use secrets that expire, and rotate them on a scheduled basis. When it is time to switch to a new secret, it is best to create a new secret, switch the application to use the new secret, and then delete the old secret from the client once it is no longer in use.

Author: ```suggestion
1. Use secrets that expire and rotate them on a schedule. When it is time to switch to a new secret, OSIsoft recommends that you create the new secret, redirect the application to use the new secret, and only delete the old secret from the client when it is no longer being used.
```

### Creating a client-credentials client

> Find the Clients page in the OSIsoft Cloud Services portal [here](https://cloud.osisoft.com/users).

Author: ```suggestion
1. Click the menu icon and click **Users** (under Security).
```

> 1. From the **Client Type:** dropdown, select **Client-Credentials**. (This is also the default selected option)

Author: ```suggestion
1. From the **Client Type:** dropdown, select **Client-Credentials**. (This is the option that is displayed, by default.)
```

1. In the toolbar, click **Add Client**.

> 1. Enter a **Name** for the client to indicate what device or application it will be used by.

Author: ```suggestion
1. Enter a **Name** for the client. 
    Use this name to identify the device or application that will use this client.
```

> 1. Enable the appropriate roles for the client. By default the client will have the Account Member role, which cannot be removed. Roles can still be modified after the client is created.

Author: ```suggestion
1. Select the roles that are appropriate for the client. 
    By default, the client has the Account Member role which cannot be removed. Roles can be modified after the client is created.
```

> 1. Optionally, set the **Token Lifetime**. This represents how long the access token will function before it expires. The default (and maximum) value is 3600 seconds (one hour). The minimum value is 60 seconds (one minute).

Author: ```suggestion
1. (Optional) Set the **Token Lifetime**. 
    Token Lifetime represents the length of time (in seconds) the access token functions before it expires. The default, 3600 seconds (one hour), is the maximum length of time. The minimum value is 60 seconds.
```

1. Click **Continue**.

> 1. You are now prompted to add a secret for the client. Optionally, you can enter a **Description** for the client secret.

Author: ```suggestion
1. In the Add Secret window, enter a **Description** for the client secret. 
```

> 1. Ensure that the expiration date is valid for the secret. By default, the secret expires after one year. It is also possible to specify that the secret **Never Expires**. Note that even in this case the secret can be manually deleted from the client later.

Author: ```suggestion
1. Enter an date that the secret expires.
    Ensure that the expiration date is valid for the secret. By default, the secret is set to expires one year after it is created. You may specify that the secret does not expire by selecting **Never Expires** checkbox. Secrets can be deleted later, including secrets that are set to never expire.
```

> 1. Click **Add**. You will be prompted with a dialog listing the **Client Id** and **Client Secret**. **IMPORTANT**: The **Client Secret** will no longer be accessible once you close the dialog. Use the **Copy** buttons to the right of these fields to save the values.

Author: ```suggestion
1. Click **Add**. 
    A window opens with the **Client Id** and **Client Secret**. 
    **IMPORTANT**: 
    Use the **Copy** buttons to the right of these fields to save the client secret and client Id in a secure location. Once the window is closed, the **Client Secret** cannot be accessed or retrieved. 
```
Do they need to save the Client Id as well?

> 1. Click **Close**. Note that **Client Id** is still visible, and if you forgot to save the **Client Secret**, you can select the client from the list and use the **Details** tab to remove the old secret and add a new secret.

Author: ```suggestion
1. Click **Close**. 
    Note: In the list of secrets, the **Client Id** is still visible. If you did not save the **Client Secret**, select the client in the list and use the **Details** pane to remove the old secret and add a new secret.
```

> ## Authorization Code

Author: ```suggestion
## Authorization code clients
```

> Authorization Code Clients are used for customer web applications that use OCS as their backend. They provide a secure means of authenticating users of the website to view OCS assets. Creating an Authorization Code Client results in the generation of a Client ID, which must be specified in the web application using the client to authenticate users.

Author: ```suggestion
Authorization code clients are used with customer web applications that use OCS as their backend. They provide a secure means of authenticating users of the website to view OCS assets. The authorization code client is paired with a client ID. The web application that is using the client to authenticate users must include the client ID in its code.
```

> Authorization code clients are used to authenticate using any browser, and an authorization code is provided to the client upon successful authentication. The authorization code is exchanged for an access token using PKCE (Proof Code for Code Exchange), which is a more secure authentication flow. No refresh token is provided.

Author: ```suggestion
Authorization code clients are used to authenticate using any browser. Upon successful authentication, an authorization code is provided to the client. This authorization code is exchanged for an access token using PKCE (Proof Code for Code Exchange) which is a more secure authentication flow. No refresh token is provided.
```

### Authorization Code PI Core Counterpart

> Authorization Code clients have no direct equivalent, but are conceptually similar to Trusts in the PI Data Archive since they are used to allow users to access OCS if they meet certain criteria. However, instead of allowing access via application name and/or IP Address, Authorization Code clients instead allow users to log in using their own credentials from applications that run at certain URLs.

Reviewer 2: Not sure about this.  I think of them more like mapping.  Since the person running the application is given the security they are supposed to get and not just the security to the app itself

Reviewer: See if you like this better... it has elements of mappings and trusts combined.

> Authorization Code clients have no direct equivalent, but are similar to a combination of a Trust and Mappings in the PI Data Archive. These clients are similar to Trusts since they only allow users to access OCS if the application using them meets certain criteria, such as being served at a specific URL. However, like a mapping, they also require authentication as a known user account within the tenant.

Author: ```suggestion
Authorization code clients have no direct PI Core equivalent, but they are similar to the combined behavior of a trust and mappings in PI Data Archive. These clients are similar to trusts because they only allow users to access OCS if the application that uses them meets certain criteria, for example, the application must be served at a specific URL. However, like a mapping, authorization code clients must be authenticated as a known user account within the tenant.
```

Reviewer: Where you adjusted to `authorization code clients must be authenticated` I would adjust this further to `authorization code clients require the user to authenticate`. It's not the authorization code client that needs to authenticate, it's the user that's logging in using one of these clients.

Author: Thanks. Done. Committed changes.

### Authorization code best practices

> 1. Use Authorization Code clients in web applications or services where users need to authenticate and it is not possible to store a client secret securely.

Author: ```suggestion
1. Use authorization code clients in web applications or with services where users must be authenticated, and it is not possible to store a client secret securely.
```

> 1. Use authorization code clients in web applications or with services where users must be authenticated, and it is not possible to store a client secret securely.

Author: ```suggestion
1. Use authorization code clients in web applications or with services where users must be authenticated and it is not possible to store a client secret securely.
```

> 1. Since a refresh token is not provided in this flow, web applications should use an iframe to request a new token before the token expires. Otherwise, the user will have to explicitly log in again to get a new token once their token expires.

Author: ```suggestion
1. Because refresh tokens are not generated in this flow, web applications should use an iframe to request a new token before the existing token expires. Otherwise, the user will have to explicitly log in again to get a new token once their token expires.
```

Author: What is an iframe? When I google it, it brings up HTML iframe. I guess they'll know what this is? Can we say "HTML iframe"?

### Creating an authorization code client

> Find the Clients page in the OSIsoft Cloud Services portal [here](https://cloud.osisoft.com/users).

Author: ```suggestion
1. Click the menu icon and click **Users** (under Security).
```

Reviewer: Looks like this should be `click **Clients**`

Author: Okay, thanks. Done.

Reviewer: @lasato-OSI Looks like this can no longer be committed because it is outdated (I think another suggestion completely removed this line).

Reviewer: Got this committed directly instead since the suggestion doesn't work now.

> 1. From the **Client Type:** dropdown, select **Authorization Code**.

Author: ```suggestion
1. From the **Client Type:** dropdown list, select **Authorization Code**.
```

1. In the toolbar, click **Add Client**.

> 1. Enter a **Name** for the client to indicate what application it will be used by.

Author: ```suggestion
1. Enter a **Name** for the client. 
    Use this name to identify the device or application that will use this client.
```

> 1. Enter the **Allowed Redirect URL(s)** for the application. This will be specified by the application when authenticating against OCS, and the OCS identity server will direct back to this URL after successful authentication.

Author: ```suggestion
1. Enter the URLs in the **Allowed Redirect URL(s)** text box.
    The application will specify the URLs in this list when it authenticates against OCS, and the OCS identity server will return the results of the authentication to this URL.
```

Reviewer: Would be more accurate to say `The application will specify one of the URLs in this list`

Author: Done.

> 1. Optionally, enter **Allowed Logout Redirect URL(s)**. Similar to normal redirect URL(s), this is specified by the application when logging out of OCS, and the OCS identity server will direct back to this URL after successful logout.

Author: ```suggestion
1. (Optional) Enter URLs in the **Allowed Logout Redirect URL(s)** text box. 
     This is similar to the **Allowed Redirect URL(s)**. The application specifies the **Allowed Logout Redirect URL(s)** when it  logs out of OCS, and the OCS identity server returns the successful log out result back to this URL
```
Is "direct back" a programming term? I was trying to find a less jargony way of expressing this.

Reviewer: Redirect is probably the correct term, but agreed. Maybe `The application specifies an **Allowed Logout Redirect URL** from this list when it logs out of OCS, and the OCS identity server sends the user to that URL after a successful log out.`?

Author: Done

Reviewer: Suggestion won't work now, but I got this committed directly.

> 1. Optionally, enter **Allowed CORS Origin(s)**. This can be used to designate other URL(s) from which the application is allowed to make requests against OCS. For example, this may be necessary if the user authenticates from an application running at `https://authenticate.com`, but is then sent to an application running at `https://application.com`.

Author: ```suggestion
1. (Optional) Enter URLs in the **Allowed CORS Origin(s)** text box. 
    This can be used to designate other URL(s) from which the application is allowed to make requests against OCS. For example, this may be necessary if the user authenticates from an application running at `https://authenticate.com`, but is the result of the authentication is sent to an application running at `https://application.com`.
```

> 1. Optionally, set the **Token Lifetime**. This represents how long the access token will function before it expires. The default (and maximum) value is 3600 seconds (one hour). The minimum value is 60 seconds (one minute).

Author: ```suggestion
1. (Optional) Set the **Token Lifetime**. 
    Token Lifetime represents the length of time (in seconds) the access token is valid before it expires. The default, 3600 seconds (one hour), is the maximum length of time. The minimum value is 60 seconds.
```

> 1. Click **Save**. Note the **Client Id** from the list, which must be specified when the application makes an authentication request. The client is now ready to use.

Author: ```suggestion
1. Click **Add**.
    The Client Successfully Created window displays the Client Id for the client.  The application must specify this Client Id when it makes an authentication request. You can also retrieve the **Client Id** from the list of clients.

## Hybrid clients

> Hybrid clients are used by native and server-side web applications where you authenticate using any browser, and an access token is retrieved using the server-side code. A refresh token can also be provided.

Author: ```suggestion
Hybrid clients are used by native and server-side web applications. Authentication can be performed using any browser. The server-side code retrieves an access token and a refresh token can also be provided.
```

### Hybrid PI Core Counterpart

> Hybrid clients have no direct equivalent, but are similar to Trusts in the PI Data Archive since they are used to allow users to access OCS if they meet certain criteria. However, instead of allowing access via application name and/or IP Address, Hybrid clients instead allow access users to log in using their own credentials from applications that use a secure backend.

Reviewer 2: Not sure about this.  I think of them more like mapping.  Since the person running the application is given the security they are supposed to get and not just the security to the app itself

Reviewer: See if you like this better... it has elements of mappings and trusts combined.

> Hybrid clients have no direct equivalent, but are similar to a combination of a Trust and Mappings in the PI Data Archive. These clients are similar to Trusts since they only allow users to access OCS if the application using them meets certain criteria, such as being served at a specific URL. However, like a mapping, they also require authentication as a known user account within the tenant.

Author: ```suggestion
Hybrid clients have no direct PI Core equivalent, but they are similar to the combined behavior of a trust and mappings in PI Data Archive. These clients are similar to trusts because they only allow users to access OCS if the application that uses them meets certain criteria, for example, the application must be served at a specific URL. However, like a mapping, hybrid clients must be authenticated as a known user account within the tenant.```

Reviewer: Similar to above, where you adjusted to `hybrid clients must be authenticated` I would adjust this further to `hybrid clients require the user to authenticate`.

Author: Done. Committed changes.

### Hybrid best practices

> 1. Use Hybrid clients in web applications or services where users can use a browser to authenticate against OCS, but a secure backend that can store a secret is used to perform the actual authentication process.

Author: ```suggestion
1. Use hybrid clients in web applications or services where users authenticate against OCS through a web browser, but a secure backend that stores the secrets performs the actual authentication.
```

> 1. Use caution when determining whether to allow a refresh token for your Hybrid client. If possible, it is more secure to use an iframe to request a new token before the old token expires.

Author: ```suggestion
1. Use caution when deciding whether to allow a refresh token for your hybrid client. Where it is possible, it is a more secure practice to use an iframe to request a new token before the old token expires rather than using a refresh token.
```

> ### Creating a hybrid vlient

Author: ```suggestion
### Creating a hybrid client
```

> Find the Clients page in the OSIsoft Cloud Services portal [here](https://cloud.osisoft.com/users).

Author: ```suggestion
1. Click the menu icon and click **Users** (under Security).
```

> 1. From the **Client Type:** dropdown, select **Hybrid**.

Author: ```suggestion
1. From the **Client Type:** dropdown list, select **Hybrid**.
```

1. In the toolbar, click **Add Client**.

> 1. Enter a **Name** for the client to indicate what application it will be used by.

Author: ```suggestion
1. Enter a **Name** for the client. 
    Use this name to identify the application that will use this client.

```

> 1. Optionally, check the box **Allow Refresh Token** if the application will use refresh tokens to keep users logged in to OCS.

Author: ```suggestion
1. (Optional) Select the **Allow Refresh Token** check box if the application uses refresh tokens to keep users logged in to OCS.
```

> 1. Enter the **Allowed Redirect URL(s)** for the application. This will be specified by the application when authenticating against OCS, and the OCS identity server will direct back to this URL after successful authentication.

Author: ```suggestion
1. Enter the URLs in the **Allowed Redirect URL(s)** text box.
    The application will specify the URLs in this list when it authenticates against OCS, and the OCS identity server will return the results of the authentication to this URL.

```

Reviewer: It's more accurate to say `the application will specify one of the URLs in this list`

Author: Done

> 1. Optionally, enter **Allowed Logout Redirect URL(s)**. Similar to normal redirect URL(s), this is specified by the application when logging out of OCS, and the OCS identity server will direct back to this URL after successful logout.

Author: ```suggestion
1. (Optional) Enter URLs in the **Allowed Logout Redirect URL(s)** text box. 
     This is similar to the **Allowed Redirect URL(s)**. The application specifies the **Allowed Logout Redirect URL(s)** when it  logs out of OCS, and the OCS identity server returns the successful log out result back to this URL
```

Reviewer: I'd suggest instead `The application specifies one of the **Allowed Logout Redirect URL(s)** from this list when it logs a user out of OCS, and the OCS identity server sends the user to that URL after successful log out.`

Author: Done

> 1. Optionally, set the **Token Lifetime**. This represents how long the access token will function before it expires. The default (and maximum) value is 3600 seconds (one hour). The minimum value is 60 seconds (one minute).

Author: ```suggestion
1. (Optional) Set the **Token Lifetime**. 
    Token Lifetime represents the length of time (in seconds) the access token is valid before it expires. The default, 3600 seconds (one hour), is the maximum length of time. The minimum value is 60 seconds.
```

> 1. Click **Save**. Note the **Client Id** from the list, which must be specified when the application makes an authentication request. The client is now ready to use.

Author: ```suggestion
1. Click **Add**.
    The Client Successfully Created window displays the Client Id for the client.  The application must specify this Client Id when it makes an authentication request. After you close this window, you can also retrieve the **Client Id** from the list of clients.
```

# Clients

## Definition

Clients provide a way for users or applications to authenticate against OSIsoft Cloud Services (OCS) from applications other than the OCS portal. OCS supports three types of clients, each of which support different types of applications.

You must have the **Account Administrator** role to add and manage clients in a tenant.

## Client-credentials clients

Client-credentials clients are used for server-to-server communication where no user interaction is required. The client typically authenticates with the token endpoint using its client ID and secret. A secret is a unique key generated for each client to connect to OSIsoft assets, resources, and services for a time-limited period.

### Client-credentials PI Core counterpart

> Client-credentials clients are very similar to Windows service accounts that might be used to authenticate against PI Data Archive or PI AF server. Therefore, it is very important to keep secrets secure in the same way that it is important to keep service account passwords secure.

Author: ```suggestion
Client-credentials clients are very similar to Microsoft Windows service accounts that might be used to authenticate against PI Data Archive or PI AF server. Therefore, it is very important to keep secrets secure in the same way that it is important to keep service account passwords secure.
```

### Client-credentials best practices

1. Create a separate client-credentials client for each device or instance of an application that connects to OCS. This ensures that secrets can be discretely managed for individual applications and that you know which applications are connecting to OCS.

1. Ensure that client secrets are stored securely where they are used.

1. Use secrets that expire and rotate them on a schedule. When it is time to switch to a new secret, OSIsoft recommends that you create the new secret, redirect the application to use the new secret, and only delete the old secret from the client when it is no longer being used.

### Creating a client-credentials client

1. Click the menu icon and click **Users** (under Security).

> 1. From the **Client Type:** dropdown list, select **Client-Credentials**. (This is the option that is displayed, by default.)

Author: ```suggestion
1. From the **Client Type** dropdown list, select **Client-Credentials**. (This is the option that is displayed, by default.)
```

1. In the toolbar, click **Add Client**.

1. Enter a **Name** for the client. 
    Use this name to identify the device or application that will use this client.

1. Select the roles that are appropriate for the client. 
    By default, the client has the Account Member role which cannot be removed. Roles can be modified after the client is created.

1. (Optional) Set the **Token Lifetime**. 
    Token Lifetime represents the length of time (in seconds) the access token functions before it expires. The default, 3600 seconds (one hour), is the maximum length of time. The minimum value is 60 seconds.

1. Click **Continue**.

1. In the Add Secret window, enter a **Description** for the client secret. 

1. Enter a date that the secret expires.
    Ensure that the expiration date is valid for the secret. By default, the secret is set to expire one year after creation. You may specify that the secret does not expire by selecting the **Never Expires** checkbox. Secrets can be deleted later, including secrets that are set to never expire.

1. Click **Add**. 
    A window opens with the **Client Id** and **Client Secret**. 
    **IMPORTANT**: 

>     Use the **Copy** buttons to the right of these fields to save the client secret and client Id. You will need to reference these in your applications. Once the window is closed, the **Client Secret** cannot be accessed or retrieved. 

Author: ```suggestion
    Use the **Copy** buttons to the right of these fields to save the client secret and client ID. You will need to reference these in your applications. Once the window is closed, the **Client Secret** cannot be accessed or retrieved. 
```

1. Click **Close**. 
    Note: In the list of secrets, the **Client Id** is still visible. If you did not save the **Client Secret**, select the client in the list and use the **Details** pane to remove the old secret and add a new secret.

## Authorization code clients

Authorization code clients are used with customer web applications that use OCS as their backend. They provide a secure means of authenticating users of the website to view OCS assets. The authorization code client is paired with a client ID. The web application that is using the client to authenticate users must include the client ID in its code.

Authorization code clients are used to authenticate using any browser. Upon successful authentication, an authorization code is provided to the client. This authorization code is exchanged for an access token using PKCE (Proof Code for Code Exchange) which is a more secure authentication flow. No refresh token is provided.

> ### Authorization code PI Core counterpart

Author: ```suggestion
### Authorization code client PI Core counterpart
```

Authorization code clients have no direct PI Core equivalent, but they are similar to the combined behavior of a trust and mappings in PI Data Archive. These clients are similar to trusts because they only allow users to access OCS if the application that uses them meets certain criteria, for example, the application must be served at a specific URL. However, like a mapping, authorization code clients require the user to authenticate as a known user account within the tenant.

> ### Authorization code best practices

Author: ```suggestion
### Authorization code client best practices
```

1. Use authorization code clients in web applications or with services where users must be authenticated and it is not possible to store a client secret securely.

1. Because refresh tokens are not generated in this flow, web applications should use an iframe to request a new token before the existing token expires. Otherwise, the user will have to explicitly log in again to get a new token once their token expires.

### Creating an authorization code client

> Find the Clients page in the OSIsoft Cloud Services portal [here](https://cloud.osisoft.com/users).

Author: ```suggestion
```

> 1. From the **Client Type:** dropdown list, select **Authorization Code**.

Author: ```suggestion
1. From the **Client Type** dropdown list, select **Authorization Code**.
```

1. In the toolbar, click **Add Client**.

1. Enter a **Name** for the client. 
    Use this name to identify the device or application that will use this client.

1. Enter the **Allowed Redirect URL(s)** for the application. This will be specified by the application when authenticating against OCS, and the OCS identity server will direct back to this URL after successful authentication.

> 1. Optionally, enter **Allowed Logout Redirect URL(s)**. Similar to normal redirect URL(s), this is specified by the application when logging out of OCS, and the OCS identity server will direct back to this URL after successful logout.

Author: ```suggestion
1. (Optional) Enter **Allowed Logout Redirect URL(s)**. Similar to normal redirect URL(s), this is specified by the application when logging out of OCS, and the OCS identity server will direct back to this URL after successfully logging out.
```

1. (Optional) Enter URLs in the **Allowed CORS Origin(s)** text box. 

>     This can be used to designate other URL(s) from which the application is allowed to make requests against OCS. For example, this may be necessary if the user authenticates from an application running at `https://authenticate.com`, but the result of the authentication is sent to an application running at `https://application.com`.

Author: ```suggestion
    This can be used to designate other URL(s) from which the application is allowed to make requests against OCS. For example, this may be necessary if the user authenticates from an application running at `https://authenticate.com`, but the result of the authentication is sent to an application running at a different URL, for example,  `https://application.com`.
```

1. (Optional) Set the **Token Lifetime**. 
    Token Lifetime represents the length of time (in seconds) the access token is valid before it expires. The default, 3600 seconds (one hour), is the maximum length of time. The minimum value is 60 seconds.

1. Click **Add**.
    The Client Successfully Created window displays the Client Id for the client.  The application must specify this Client Id when it makes an authentication request. After you close this window, you can also retrieve the **Client Id** from the list of clients.

## Hybrid clients

Hybrid clients are used by native and server-side web applications. Authentication can be performed using any browser. The server-side code retrieves an access token and a refresh token can also be provided.

> ### Hybrid PI Core counterpart

Author: ```suggestion
### Hybrid client PI Core counterpart
```

> Hybrid clients have no direct PI Core equivalent, but they are similar to the combined behavior of a trust and mappings in PI Data Archive. These clients are similar to trusts because they only allow users to access OCS if the application that uses them meets certain criteria, for example, the application must be served at a specific URL. However, like a mapping, hybrid clients require the user to authenticate as a known user account within the tenant.```

Author: ```suggestion
Hybrid clients have no direct PI Core equivalent, but they are similar to the combined behavior of a trust and mappings in PI Data Archive. These clients are similar to trusts because they only allow users to access OCS if the application that uses them meets certain criteria, for example, the application must be served at a specific URL. However, like a mapping, hybrid clients require the user to authenticate as a known user account within the tenant.
```

> ### Hybrid best practices

Author: ```suggestion
### Hybrid client best practices
```

1. Use hybrid clients in web applications or services where users authenticate against OCS through a web browser, but a secure backend that stores the secrets performs the actual authentication.

1. Use caution when deciding whether to allow refresh tokens for your hybrid client. Where possible, it is a more secure practice to use an iframe to request a new token before the old token expires rather than use a refresh token.

### Creating a hybrid client

1. Click the menu icon and click **Users** (under Security).

> 1. From the **Client Type:** dropdown list, select **Hybrid**.

Author: ```suggestion
1. From the **Client Type** dropdown list, select **Hybrid**.
```
